packages/http-server/src/http-server.ts

```diff
@@ -0,0 +1,114 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/http-server
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {createServer, Server, ServerRequest, ServerResponse} from 'http';
+import {AddressInfo} from 'net';
+import * as pEvent from 'p-event';
+
+export type HttpRequestListener = (
+  req: ServerRequest,
+  res: ServerResponse,
+) => void;
+
+/**
+ * Object for specifyig the HTTP / HTTPS server options
+ */
+export type HttpServerOptions = {
+  port?: number;
+  host?: string;
+  protocol?: HttpProtocol;
+};
+
+export type HttpProtocol = 'http' | 'https'; // Will be extended to `http2` in the future
+
+/**
+ * HTTP / HTTPS server used by LoopBack's RestServer
+ *
+ * @export
+ * @class HttpServer
+ */
+export class HttpServer {
+  private _port: number;
+  private _host?: string;
+  /**
+   * Protocol, default to `http`
+   */
+  private _protocol: HttpProtocol;
+  private _address: AddressInfo;
+  private httpRequestListener: HttpRequestListener;
+  private httpServer: Server;
+
+  /**
+   * @param httpServerOptions
+   * @param httpRequestListener
+   */
+  constructor(
+    httpRequestListener: HttpRequestListener,
+    httpServerOptions?: HttpServerOptions,
+  ) {
+    this.httpRequestListener = httpRequestListener;
+    if (!httpServerOptions) httpServerOptions = {};
+    this._port = httpServerOptions.port || 0;
+    this._host = httpServerOptions.host || undefined;
+    this._protocol = httpServerOptions.protocol || 'http';
+  }
+
+  /**
+   * Starts the HTTP / HTTPS server
+   */
+  public async start() {
+    this.httpServer = createServer(this.httpRequestListener);
+    this.httpServer.listen(this._port, this._host);
+    await pEvent(this.httpServer, 'listening');
```

Comment:
I wonder why we don't use a callback for `httpServer.listen()` and `close()`.

Reply:
Good idea, will add.

Reply:
Please note that `pEvent` will listen for `error` events too. As a result, `server.start()` returns a rejected promise if the server cannot be started.

The callback for `httpServer.listen` is not error-first, it's just a regular event callback for `listening` event. See https://nodejs.org/dist/latest-v10.x/docs/api/net.html#net_server_listen:

> When the server starts listening, the 'listening' event will be emitted. The last parameter callback will be added as a listener for the 'listening' event.

@hacksparrow Please start by adding a test to verify that `start()` returns a rejected promise when the server cannot be started. Quoting from the same doc page:

> One of the most common errors raised when listening is `EADDRINUSE`. This happens when another server is already listening on the requested port / path / handle.

In the test:
 - create and start a server listening on port `0` (let the OS pick an available port)
 - create another server, set the port number explicitly to the same value as reported by the server started in the previous step
 - verify that `start()` returns a rejected promise and the error has `code` set to `EADDRINUSE` - see http://shouldjs.github.io/#assertion-rejectedwith

Once we have that test in place, you can try to use the promisifed `listen` method and see how easy/difficult it is to correctly handle errors too (compared to what we have now).

Reply:
@bajtos added the test.

@raymondfeng why have callbacks, when we can just `await`?